rfcs/0038-rosetta-construction-api.md

```diff
@@ -0,0 +1,424 @@
+
+#### Operations docs
+
+[operations-docs]: #operations-docs
+
+Add examples of each kind of transaction that one may want to construct as JSON files. Eventually we'd want one for each type of transaction, but for now it suffices to just include a payment.
+
+For example: The following expressrion would be saved in `payment.json`
+
+```
+[{
+  "operation_identifier": ...,
+  "amount": ...,
+  "type": "Payment_source_dec"
+},
+{
+  "operation_identifier": ...,
+  "amount": ...,
+  "type": "Payment_receiver_inc"
+},
+...
+]
+```
+
+This is useful for manual testing purposes and sets us up for integration with the [construction-portion of rosetta-cli integration](https://community.rosetta-api.org/t/feedback-request-automated-construction-api-testing-improvements/146/4).
+
+#### Inverted operations map
+
+[inverted-operations-map]: #inverted-operations-map
+
+[via Preprocess](#preprocess)
+
+Write a function that recovers the transactions that are associated with some set of operations. We should create a test `forall (t : Transaction). op^-1(op(t)) ~= t` which enumerates all the kinds of transactions. For an initial release it suffices to test this for payments.
+
+#### Preprocess Endpoint
+
+[preprocess-endpoint]: #preprocess-endpoint
+
+[via Preprocess](#preprocess)
+
+First [invert the operations](#inverted-operations-map) into a transaction, we find the sender and include the public key in the response. The options type will be defined as follows:
+
+```ocaml
+module Options = struct
+  type t =
+    { sender : string (* base58-ecoded compressed public key *)
+    }
+    [@@deriving yojson]
+end
+```
+
+#### Metadata Endpoint
+
+[preprocess-endpoint]: #preprocess-endpoint
+
+[via Metadata](#metadata)
+
+This is a simple GraphQL query. This endpoint should be easy to implement.
+
+#### Unsigned transaction encoding
+
+[encoded-unsigned-transaction]: #encoded-unsigned-transaction
+
+[via Payloads](#payloads)
+
+The Rosetta spec leaves the encoding of unsigned transactions implementation-defined. Since we'll default to using our client-sdk's signing mechanism our encoding will be precisely the JSON input (stringified) demanded by the "unsafe" method of the client-sdk API:
+
+```reasonml
+// Taken from Client-SDK code
+
+type stakeDelegation = {
+  [@bs.as "to"]
+  to_: publicKey,
+  from: publicKey,
+  fee: uint64,
+  nonce: uint32,
+  memo: option(string),
+  validUntil: option(uint32),
+};
+
+type payment = {
+  [@bs.as "to"]
+  to_: publicKey,
+  from: publicKey,
+  fee: uint64,
+  amount: uint64,
+  nonce: uint32,
+  memo: option(string),
+  validUntil: option(uint32),
+};
+
+type transaction = stakeDelegation | payment
+```
+
+Note that our client-sdk only has support for signing payments and delegations but this version of the construction API only supports those transactions types as well.
```

Comment:
above it says only payments supported, should it mention delegations?

Reply:
I intended for this doc to explain payments+delegations (assuming payments may be implemented first in its entirety, and then delegations to be implemented)

Reply:
Please let me know if there is some part of the doc that doesn't show this